ic-utils/src/interfaces/management_canister/builders.rs

```diff
@@ -507,6 +521,280 @@ impl<'agent, 'canister: 'agent> AsyncCall<()> for InstallCodeBuilder<'agent, 'ca
+    canister_id: Principal,
+    // more precise lifetimes are used here at risk of annoying the user
+    // because `wasm` may be memory-mapped which is tricky to lifetime
+    wasm: &'builder [u8],
+    arg: Argument,
+    mode: InstallMode,
+}
+
+impl<'agent: 'canister, 'canister: 'builder, 'builder> InstallBuilder<'agent, 'canister, 'builder> {
+    // Messages are a maximum of 2MiB. Thus basic installation should cap the wasm and arg size at 1.85MiB, since
+    // the current API is definitely not going to produce 150KiB of framing data for it.
+    const CHUNK_CUTOFF: usize = (1.85 * 1024. * 1024.) as usize;
+
+    /// Create a canister installation builder.
+    pub fn builder(
+        canister: &'canister ManagementCanister<'agent>,
+        canister_id: &Principal,
+        wasm: &'builder [u8],
+    ) -> Self {
+        Self {
+            canister,
+            canister_id: *canister_id,
+            wasm,
+            arg: Default::default(),
+            mode: InstallMode::Install,
+        }
+    }
+
+    /// Set the argument to the installation, which will be passed to the init
+    /// method of the canister. Can be called at most once.
+    pub fn with_arg<Argument: CandidType>(mut self, arg: Argument) -> Self {
+        self.arg.set_idl_arg(arg);
+        self
+    }
+    /// Set the argument with multiple arguments as tuple to the installation,
+    /// which will be passed to the init method of the canister. Can be called at most once.
+    pub fn with_args(mut self, tuple: impl ArgumentEncoder) -> Self {
+        assert!(self.arg.0.is_none(), "argument is being set more than once");
+        self.arg = Argument::from_candid(tuple);
+        self
+    }
+    /// Set the argument passed in to the canister with raw bytes. Can be called at most once.
+    pub fn with_raw_arg(mut self, arg: Vec<u8>) -> Self {
+        self.arg.set_raw_arg(arg);
+        self
+    }
+
+    /// Pass in the [InstallMode].
+    pub fn with_mode(self, mode: InstallMode) -> Self {
+        Self { mode, ..self }
+    }
+
+    /// Invoke the installation process. This may result in many calls which may take several seconds;
+    /// use [`call_and_wait_with_progress`](Self::call_and_wait_with_progress) if you want progress reporting.
+    pub async fn call_and_wait(self) -> Result<(), AgentError> {
+        self.call_and_wait_with_progress()
+            .try_for_each(|_| ready(Ok(())))
+            .await
+    }
+
+    /// Invoke the installation process. The returned stream must be awaited to completion; it is used to track progress,
+    /// as installation may take arbitrarily long, and is intended to be passed to functions like `indicatif::ProgressBar::wrap_stream`.
+    /// There are exactly [`size_hint().0`](Stream::size_hint) steps.
+    pub fn call_and_wait_with_progress(
+        self,
+    ) -> impl Stream<Item = Result<(), AgentError>> + 'builder {
+        let stream_res = /* try { */ (move || {
```

Comment:
what is going on with this?
> /* try { */

Reply:
An immediately-invoked closure is being used to emulate a try block (local scope for `?`), which does not yet exist.